### PR TITLE
common/options: convert some TYPE_[U]INT -> TYPE_SIZE

### PR DESCRIPTION
**Pull request by liewegas**
The practical difference is now the K, M, etc suffixes are interpreted: for [U]INT, they are SI units, and for SIZE they are IEC base-2 (2^10 KiB GiB etc).

### CODE REVIEW

**Comment by jcsp** [comment]
There was some discussion over types vs separate units in an earlier PR for adding units (https://github.com/ceph/ceph/pull/20163), which was superseded by a PR adding types (https://github.com/ceph/ceph/pull/20419)

**Comment by jcsp** [comment]
I'm not generally a fan of having types that differ only by a factor (like TYPE_MB and TYPE_BYTES) -- the correct way to display either will be to convert it to a MB/GB/TB version that is compact enough for display.  Question is whether the consumer of the option really requires an integer number of MB, or whether the setting was just made that way arbitrarily?

**Comment by jcsp** [comment]
> We also have a lot of time interval options that are seconds. We have macros for _hr _day etc to make the source easy to read, but it woudl be great to allow the same when setting options and to format the options in a nice way when displaying.

An order-of-magnitude display hint would be a good thing (in many cases this could be inferred from the default value) -- this part should probably not be part of the type though, as options of any unit might want it to hint their order of magnitude.  For example, a bytes setting might want to hint that while 1 byte is a valid setting, the user almost certainly wants to specify metabytes.

**Comment by jcsp** [comment]
ping @LenzGr -- any thoughts on how the display hints around options should be?

**Comment by LenzGr** [comment]
Hmm. I personally am a big fan of storing such information in it's "native" type (e.g. plain bytes, seconds) and then let the tools that either display this information to or obtain it from the user take care of the appropriate transformation from or into human-readable formats (e.g. displaying bytes as GiB, converting hours/minutes or a date into seconds).

This reduces complexity when storing the information, reduces potential conversion errors when transmitting that information via several layers and provides much more flexibility when displaying it (or receiving it).

**Comment by s0nea** [comment]
I agree with @LenzGr. I discussed the dashboard perspective with @Exotelis and the best way would be to get the plain bytes. The dashboard is already capable of converting it into the actual unit (MiB, GiB, ...) itself.

**Comment by liewegas** [comment]
rebased on master.  I think the PR as is is a reasonable step, then: we just change some INT and UINT to SIZE, the effect of which is that M G etc are interpreted as base-2 instead of base-10 when parsing input.

**Comment by jcsp** [comment]
I think the existing use of TYPE_SIZE was specifically for byte quantities (we ended up with this rather than having a separate field for the unit).  So settings like xio_queue_depth, xio_portal_threads shouldn't be TYPE_SIZE unless we also add a separate unit field

File: src/common/options.cc [hunk 1]
```diff
@@ -752,47 +752,47 @@ std::vector<Option> get_global_options() {
     .set_default(false)
     .set_description(""),
 
-    Option("xio_queue_depth", Option::TYPE_INT, Option::LEVEL_ADVANCED)
+    Option("xio_queue_depth", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
```
Comment (tchaikov) [inline src/common/options.cc]: i agree with @jcsp TYPE_SIZE if for quantities like packet sizes but not for not for settings for numbers, like numbers of packets per seconds.

**Comment by liewegas** [comment]
Updated.

We have 2 options that are clearly byte size (rgw_bucket_default_quota_max_size and rgw_user_default_quota_max_size) but they default to -1 (they're signed).  Should we have a new TYPE_SSIZE for that?

**Comment by tchaikov** [comment]
can we reserve `std::numeric_limits<size_t>::max()` for the `-1` case? or this idea is just crazy?

File: src/common/options.cc [hunk 2]
```diff
@@ -3579,23 +3579,23 @@ std::vector<Option> get_global_options() {
...
+    Option("osd_pg_delete_cost", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
     .set_default(1<<20)
     .set_description(""),
 
     Option("osd_scrub_priority", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
     .set_default(5)
     .set_description("Priority for scrub operations in work queue"),
 
-    Option("osd_scrub_cost", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    Option("osd_scrub_cost", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
```
Comment (tchaikov) [inline src/common/options.cc]: IMHO, this is not a size.

File: src/common/options.cc [hunk 3]
```diff
@@ -3607,7 +3607,7 @@ std::vector<Option> get_global_options() {
     .set_default(5)
     .set_description(""),
 
-    Option("osd_recovery_cost", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    Option("osd_recovery_cost", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
```
Comment (tchaikov) [inline src/common/options.cc]: not a size.

File: src/common/options.cc [hunk 4]
```diff
@@ -4727,7 +4727,7 @@ std::vector<Option> get_global_options() {
     .set_default(4_K)
     .set_description(""),
 
-    Option("journal_max_corrupt_search", Option::TYPE_UINT, Option::LEVEL_DEV)
+    Option("journal_max_corrupt_search", Option::TYPE_SIZE, Option::LEVEL_DEV)
```
Comment (tchaikov) [inline src/common/options.cc]: could take this opportunity to remove this unused option?

File: src/common/options.cc [hunk 5]
```diff
@@ -5007,7 +5007,7 @@ std::vector<Option> get_rgw_options() {
     .set_default(0)
     .set_description("The maximum length of metadata value. 0 skips the check"),
 
-    Option("rgw_max_attr_name_len", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    Option("rgw_max_attr_name_len", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
```
Comment (tchaikov) [inline src/common/options.cc]: it's the length of rgw attribute name, so i don't think it's a size per se.

File: src/common/options.cc [hunk 6]
```diff
@@ -7011,7 +7011,7 @@ std::vector<Option> get_mds_options() {
     .set_default(0)
     .set_description("minimum uid of client to perform snapshots"),
 
-    Option("mds_snap_max_uid", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    Option("mds_snap_max_uid", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
```
Comment (tchaikov) [inline src/common/options.cc]: the uid is not a size. neither is its upper bound.

File: src/common/options.cc (hunk 5 above)
Comment (liewegas) [reply src/common/options.cc]: it's measuring characters/bytes? length of a string?  seems like a size to me

**Comment by liewegas** [comment]
Updated.  I got those _cost options confused with the priorityqueue costs, which are in units of bytes.

File: src/common/options.cc (hunk 5 above)
Comment (tchaikov) [reply src/common/options.cc]: it's measuring number of characters.

File: src/common/options.cc [hunk 7]
```diff
@@ -3017,23 +3017,23 @@ std::vector<Option> get_global_options() {
...
+    Option("osd_copyfrom_max_chunk", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
     .set_default(8_M)
     .set_description(""),
 
     Option("osd_push_per_object_cost", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
     .set_default(1000)
     .set_description(""),
 
-    Option("osd_max_push_cost", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    Option("osd_max_push_cost", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
```
Comment (tchaikov) [inline src/common/options.cc]: it's not a size, see also `osd_push_per_object_cost`.

File: src/common/options.cc [hunk 8]
```diff
@@ -3579,15 +3579,15 @@ std::vector<Option> get_global_options() {
     .set_default(5)
     .set_description(""),
 
-    Option("osd_snap_trim_cost", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    Option("osd_snap_trim_cost", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
```
Comment (tchaikov) [inline src/common/options.cc]: not a size.

File: src/common/options.cc [hunk 9]
```diff
@@ -3579,15 +3579,15 @@ std::vector<Option> get_global_options() {
...
     .set_default(1<<20)
     .set_description(""),
 
     Option("osd_pg_delete_priority", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
     .set_default(5)
     .set_description(""),
 
-    Option("osd_pg_delete_cost", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    Option("osd_pg_delete_cost", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
```
Comment (tchaikov) [inline src/common/options.cc]: not a size.

**Comment by jcsp** [comment]
+1 to @tchaikov 's suggestion to avoid negative byte counts by using `max` instead.

We probably also need to express the nullable nature of the field explicitly as well, otherwise a UI doesn't know what to do when it sees the magic "ignore this setting" value.  So maybe we do end up with a different type, but it's a "NULLABLESIZE" instead of a signed size?

**Comment by liewegas** [comment]
It seems like a FLAG_OPTIONAL is what we're after?  That would turn it into a boost::optional<foo> maybe?

The trick there is that  we need to decode how config values are parsed or stringified, since in the end these are all stored/set as strings.  Maybe "none" (and "-1" for the case of TYPE_SIZE) both mean "not set"?

Leaving this off for now since it's a bigger decision/change.

File: src/common/options.cc (hunk 9 above)
Comment (liewegas) [reply src/common/options.cc]: These costs are in units of bytes.  We map all ops to bytes units for comparison...

**Comment by liewegas** [comment]
Verified that all of the *_cost metrics are TYPE_SIZE so that it is consistent and noted it in the commit message.  We did this so that users adjusting these options had an intuitive sense of what the cost means (e.g., a cost of 1MB for delete means it is as expensive as a 1MB write).

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
@tchaikov why was the struct size_t specialization needed in options.h?  It is breaking the call size that do get_val<size_t> because std::size_t does not match the boost::variant's struct size_t.

**Comment by tchaikov** [comment]
@liewegas i agree that's awkward and confusing. because we need to differentiate `size_t` from `uint64_t`, we need a different type than `std::size_t` in `Option::value_t`.

**Pull request merged**